pkg/schema/schema.go

```diff
@@ -98,7 +98,6 @@ type User struct {
 	NoLogInit         bool     `yaml:"no_log_init,omitempty"`
 	Shell             string   `yaml:"shell,omitempty"`
 	LockPasswd        bool     `yaml:"lock_passwd,omitempty"`
-	UID               string   `yaml:"uid,omitempty"`
```

Comment:
eeek. why we can't keep a way to specify the UID? I think it's very handy in case someone wants to be really specific and make sure they have fixed uid. Also, did we check if product is depending on this?

Reply:
> yes indeed, uid/guid via the User plugin is not supported anymore as it can cause issues.

@Itxaka can you expand on this please?

Reply:
they got entities for a much more specific control over this.

IMO, it goes User (simplest, assumes things about the user creation), Entities (control over everything) and Commands (You just run the full useradd command yourself)


Keeping the UID just makes it more complex to support in all cases. Plus we have 2 different entrypoints for the same thing, makes no sense to me. In fact I would even go further and drop things like Homedir and such. Just leave user as a high approach to adding users (something like user, passwd, ssh_keys, system) and leave any more complex cases to entitites.

Reply:
> they got entities for a much more specific control over this.
> 
> IMO, it goes User (simplest, assumes things about the user creation), Entities (control over everything) and Commands (You just run the full useradd command yourself)
> 
> Keeping the UID just makes it more complex to support in all cases. Plus we have 2 different entrypoints for the same thing, makes no sense to me. In fact I would even go further and drop things like Homedir and such. Just leave user as a high approach to adding users (something like user, passwd, ssh_keys, system) and leave any more complex cases to entitites.

*Wears the community hat*

The problem I see there is a loss of UX and not being backwards compatible anymore with old cloud configs. I know entities provides much more granular and manifest-alike approach, but the user should learn how entities work first, while using the `User` plugin might be quite easy to get along with.

Reply:
> > they got entities for a much more specific control over this.
> > IMO, it goes User (simplest, assumes things about the user creation), Entities (control over everything) and Commands (You just run the full useradd command yourself)
> > Keeping the UID just makes it more complex to support in all cases. Plus we have 2 different entrypoints for the same thing, makes no sense to me. In fact I would even go further and drop things like Homedir and such. Just leave user as a high approach to adding users (something like user, passwd, ssh_keys, system) and leave any more complex cases to entitites.
> 
> _Wears the community hat_
> 
> The problem I see there is a loss of UX and not being backwards compatible anymore with old cloud configs. I know entities provides much more granular and manifest-alike approach, but the user should learn how entities work first, while using the `User` plugin might be quite easy to get along with.

_Wears common sense hat_

There is no point i having 2 plugins that do almost the same thing, which complicates code and generates bugs like this does.
Its worst for end users to have simple tools that creates bugs, than a simple tool and a complex tool that are clearly different and do different things.
Also its ok to not be backwards compatible always. We are not javascript nor we want to be javascript. Sometimes you have to draw the line and say, this is as much as we can do, otherwise you are piling up workaround on top of workarounds and creating more bugs, which in the end, its the worst thing for the user. Again, see javascript or python2 for examples on how keeping backwards compatibility always is detrimental to users.

I can restore the UID, but that means the bug will still be there and there is not much that we can do from our side, other than pile up a series of random workarounds on kairos and hope that its fixes it. We already fixed this twice, this would be the third fix, which probably wont fix it. I dunno.

Reply:
> Also its ok to not be backwards compatible always

I tend to disagree here, I would like to see a bit more mentality around honoring user config first.

> There is no point i having 2 plugins that do almost the same thing, which complicates code and generates bugs like this does.

They are two different plugins because one is meant to be cloud-init compatible, the other is a full-featured solution that exposes a completely different UX. You have to learn the UX of entities, and it requires a leap of will from the user to do that.

> Sometimes you have to draw the line and say, this is as much as we can do, otherwise you are piling up workaround on top of workarounds and creating more bugs

Why workaround? this is actually about keeping compatibility, not really workarounding any bug.

> Again, see javascript or python2 for examples on how keeping backwards compatibility always is detrimental to users.

They are both well popular exactly for that.

> I can restore the UID, but that means the bug will still be there and there is not much that we can do from our side, other than pile up a series of random workarounds on kairos and hope that its fixes it. We already fixed this twice, this would be the third fix, which probably wont fix it. I dunno.

I don't really get it. The ask here is just to keep the UID field - and if specified by the user to respect it.

Reply:
done. But this will still produces the same bug.

Create admin user called B, gets uid 1000
Create normal user called A, set uid spoecifically to 1000, get ownership of the B user files


Create user without uid, gets 1000
Create user with uid set to 1000, collision happens between the user uids

etc...

Reply:
ack, if the user starts to specify uids it should keep doing it all along the way for all the users and not mix the approaches unless they know what they are doing and assign e.g. 2001, etc.

If the user specifies a UID it still makes sense to respect that. Maybe we can print a warning that's already used, but that's low-hanging fruit